### ISSUE TITLE
Fix 3D grid making

### ISSUE THREAD

**Issue by LudwigBoess** [LudwigBoess/SPHtoGrid.jl#84]
The mapping onto 3D grids needs to be checked and fixed.

**Comment by jzuhone** [comment]
I found this issue because I was trying to use the code to create a 3D grid and I did get a crash. This is my code (sorry about the extraneous stuff):

```julia
import HDF5
using Glob
using SPHKernels
import CSV
using DataFrames
using SPHtoGrid

k = Cubic()

basedir = "/scratch7/jzuhone/tng/tng300/halos/z0"

t = DataFrame(CSV.File("$basedir/clusters.dat"; delim="\t"))

for row in eachrow(t)
    id = row["id"]
    fn = "$basedir/halo_$id.hdf5"
    HDF5.h5open(fn, "r") do f
        h = HDF5.attrs(f["Header"])["HubbleParam"]
        box_size = HDF5.attrs(f["Header"])["BoxSize"]/h
        g = f["PartType0"]
        center = [row["x"], row["y"], row["z"]]/h
        width = 5000.0
        par = mappingParameters(center=center,
                                x_size=width,
                                y_size=width,
                                z_size=width,
                                boxsize=box_size,
                                Npixels=512)
        m = g["Masses"][]*1e10/h
        pos = g["Coordinates"][]/h
        rho = g["Density"][]*1.0e10*h*h
        hsml = 2*((3.0*m./(4.0*pi*rho)).^(1/3))
        cube = sphMapping(pos, hsml, m, rho, rho; param=par, dimensions=3, kernel=k)
        write_fits_image("cube_$id.fits", cube, par)
    end
end
```

and here is the traceback:

```
ERROR: LoadError: MethodError: no method matching get_weight_per_pixel(::Float64, ::Float64, ::Float64, ::Float64, ::Int64, ::Int64, ::Float64, ::Float64, ::Float64, ::Cubic{Float64})

Closest candidates are:
  get_weight_per_pixel(::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any)
   @ SPHtoGrid ~/.julia/packages/SPHtoGrid/1UxDp/src/cic_interpolation/cic_2D.jl:1

Stacktrace:
  [1] calculate_weights(wk::Vector{Float64}, V::Vector{Float64}, iMin::Int64, iMax::Int64, jMin::Int64, jMax::Int64, kMin::Int64, kMax::Int64, x::Float64, y::Float64, z::Float64, hsml::Float64, hsml_inv::Float64, kernel::Cubic{Float64}, x_pixels::Int64, y_pixels::Int64)
    @ SPHtoGrid ~/.julia/packages/SPHtoGrid/1UxDp/src/cic_interpolation/cic_3D.jl:46
  [2] cic_mapping_3D(Pos::Matrix{Float64}, HSML::Vector{Float64}, M::Vector{Float64}, Rho::Vector{Float64}, Bin_Q::Vector{Float64}, Weights::Vector{Float64}; param::mappingParameters{Float64}, kernel::Cubic{Float64}, show_progress::Bool, calc_mean::Bool)
    @ SPHtoGrid ~/.julia/packages/SPHtoGrid/1UxDp/src/cic_interpolation/cic_3D.jl:155
  [3] sphMapping(Pos::Matrix{Float64}, HSML::Vector{Float64}, M::Vector{Float64}, Rho::Vector{Float64}, Bin_Quant::Vector{Float64}, Weights::Vector{Float64}, RM::Nothing; param::mappingParameters{Float64}, kernel::Cubic{Float64}, show_progress::Bool, parallel::Bool, reduce_image::Bool, return_both_maps::Bool, dimensions::Int64, calc_mean::Bool, stokes::Bool, sort_z::Bool)
    @ SPHtoGrid ~/.julia/packages/SPHtoGrid/1UxDp/src/cic_interpolation/cic_interpolation.jl:218
  [4] sphMapping
    @ ~/.julia/packages/SPHtoGrid/1UxDp/src/cic_interpolation/cic_interpolation.jl:35 [inlined]
  [5] (::var"#3#4"{DataFrameRow{DataFrame, DataFrames.Index}, Int64})(f::HDF5.File)
    @ Main /Volumes/scratch6/jzuhone/projects/illustris_tng/clusters/z0/make_cubes.jl:33
  [6] (::HDF5.var"#17#18"{HDF5.HDF5Context, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, var"#3#4"{DataFrameRow{DataFrame, DataFrames.Index}, Int64}, HDF5.File})()
    @ HDF5 ~/.julia/packages/HDF5/aiZLs/src/file.jl:98
  [7] task_local_storage(body::HDF5.var"#17#18"{HDF5.HDF5Context, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, var"#3#4"{DataFrameRow{DataFrame, DataFrames.Index}, Int64}, HDF5.File}, key::Symbol, val::HDF5.HDF5Context)
    @ Base ./task.jl:296
  [8] h5open(::var"#3#4"{DataFrameRow{DataFrame, DataFrames.Index}, Int64}, ::String, ::Vararg{String}; context::HDF5.HDF5Context, pv::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ HDF5 ~/.julia/packages/HDF5/aiZLs/src/file.jl:93
  [9] h5open(::Function, ::String, ::String)
    @ HDF5 ~/.julia/packages/HDF5/aiZLs/src/file.jl:91
 [10] top-level scope
    @ /Volumes/scratch6/jzuhone/projects/illustris_tng/clusters/z0/make_cubes.jl:17
in expression starting at /Volumes/scratch6/jzuhone/projects/illustris_tng/clusters/z0/make_cubes.jl:14
```

I assume that this is the expected error?

**Comment by LudwigBoess** [comment]
Hi, thanks for the detailed report!
This bug was actually introduced due to the 3D mapping being out of synch with the 2D mapping.
Last time I check the 3D mapping it looked like there was an issue with the indexing of the image tensor, so the assignment of particles to grid cells was off.
Unfortunately I don't have much time to look into the aspect of the code at the moment...

I fixed the bug you found in the `fix_3D` branch, so if you have time could you please check again if this solves your issue?

**Comment by LudwigBoess** [comment]
In my tests this has been fixed by #129 so I'm closing the issue for now. If the error does persist after all, this needs to be re-opened.
